### PR TITLE
chore(tooltips): refactor types and element interface

### PR DESCRIPTION
**Pull request by jzempel**
## Description

An improved approach for type/interface management. The arrangement in this PR will fix the `type` obfuscation seen in this doc-generated API: https://garden.zendesk.com/components/tooltip#tooltip

## Detail

Following this PR and internal RFC review, we'll add a new `docs/typescript.md` documenting the standard approach and start to pick off the non-standard definitions one package at a time.

## Checklist

- [ ] :ok_hand: ~design updates are Garden Designer approved (add the
      designer as a reviewer)~
- [x] :globe_with_meridians: demo is up-to-date (`yarn start`)
- [ ] :arrow_left: ~renders as expected with reversed (RTL) direction~
- [ ] :metal: ~renders as expected with Bedrock CSS (`?bedrock`)~
- [ ] :wheelchair: ~analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver~
- [x] :guardsman: includes new unit tests
- [ ] :memo: ~tested in Chrome, Firefox, Safari, Edge, and IE11~

### CODE REVIEW

File: packages/tooltips/src/elements/Tooltip.tsx [hunk 1]
```diff
@@ -14,46 +14,9 @@ import { useTooltip } from '@zendeskgarden/container-tooltip';
...
-  /** Displays the tooltip */
-  isVisible?: boolean;
-  /** @ignore ReactNode override */
-  children: React.ReactElement;
-  /** Defines the ref key used to position the tooltip */
-  refKey?: string;
-}
+import { getPopperPlacement, getRtlPopperPlacement } from '../utils/gardenPlacements';
+import { StyledTooltipWrapper, StyledTooltip } from '../styled';
+import { ITooltipProps, PLACEMENT, SIZE, TYPE } from '../utils/types';
```
Comment (jzempel) [inline packages/tooltips/src/elements/Tooltip.tsx:19]: In this PR's approach, the public interface pushes _out_ rather than pulling the interface _up_ from styled view elements. I believe this is the better angle, as the public interface is focal for typed prop data and documentation. And any styled interfaces should pull from the public interface as-needed.

Comment (jzempel) [reply packages/tooltips/src/elements/Tooltip.tsx:19]: For example, with the public interface pushed _out_, a styled component's interface could look something like the following without creating a circular code reference:

```
import { ITooltipProps } from '...';

interface IStyledTooltipProps {
  reusedProp: ITooltipProps['prop'];
}
```

File: packages/tooltips/src/elements/Tooltip.tsx [hunk 2]
```diff
@@ -194,24 +157,10 @@ Tooltip.propTypes = {
...
-    'start',
-    'start-top',
-    'start-bottom'
-  ]),
+  placement: PropTypes.oneOf(PLACEMENT),
   popperModifiers: PropTypes.any,
-  size: PropTypes.oneOf(['small', 'medium', 'large', 'extra-large']),
-  type: PropTypes.oneOf(['light', 'dark']),
+  size: PropTypes.oneOf(SIZE),
+  type: PropTypes.oneOf(TYPE),
```
Comment (jzempel) [inline packages/tooltips/src/elements/Tooltip.tsx:163]: Even though TypeScript helps enforce the values, all `propTypes` throughout the Garden codebase should be reusing ARRAY constants rather than maintaining separate lists.

File: packages/tooltips/src/utils/gardenPlacements.spec.ts [hunk 3]
```diff
@@ -33,17 +28,17 @@ describe('Garden Placement Utilities', () => {
         'start-bottom': 'left-end'
       };
 
-      Object.keys(GARDEN_PLACEMENTS).forEach(gardenPlacement => {
-        expect(getPopperPlacement(gardenPlacement as GARDEN_PLACEMENT)).toBe(
-          GARDEN_PLACEMENTS[gardenPlacement as GARDEN_PLACEMENT]
-        );
+      const gardenPlacements = Object.keys(GARDEN_PLACEMENTS) as GardenPlacement[];
```
Comment (jzempel) [inline packages/tooltips/src/utils/gardenPlacements.spec.ts:31]: Prevent multiple `as` type assertions whenever possible. See https://stackoverflow.com/questions/52856496/typescript-object-keys-return-string

File: packages/tooltips/src/utils/types.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  'left-end'
+] as const;
+
+export const SIZE = ['small', 'medium', 'large', 'extra-large'] as const;
+
+export const TYPE = ['light', 'dark'] as const;
+
+export type GardenPlacement = typeof PLACEMENT[number];
+
+export type PopperPlacement = typeof POPPER_PLACEMENT[number];
```
Comment (jzempel) [inline packages/tooltips/src/utils/types.ts:47]: See https://melvingeorge.me/blog/convert-array-into-string-literal-union-type-typescript for array-to-type conversion explanation.

File: packages/tooltips/src/elements/Tooltip.spec.tsx [hunk 5]
```diff
@@ -9,8 +9,8 @@ import React from 'react';
 import userEvent from '@testing-library/user-event';
 import { render, act, renderRtl } from 'garden-test-utils';
 import { DEFAULT_THEME, getColor } from '@zendeskgarden/react-theming';
-
-import { Tooltip, ITooltipProps } from './Tooltip';
+import { Tooltip } from './Tooltip';
+import { ITooltipProps } from '../utils/types';
```
[the remark this answers is not in the document]
Comment (jzempel) [reply packages/tooltips/src/elements/Tooltip.spec.tsx:13]: The argument against `/types` is that there's a pretty strong pattern of only having `/elements`, `/styles`, `/utils` at the src root. (After that, our evolutionary conventions get a bit chaotic 😅). Is `/types` a strong enough concept to pull out to the src root? And do we end up creating an unnecessary decision between `/types/{file}.ts`, `types.ts`, and `/types/index.ts`? I guess this is why I opted for `/utils/types` – because I didn't really want to think through the ramifications for every package.

File: packages/tooltips/src/elements/Tooltip.tsx (hunk 1 above)
Comment (jzempel) [reply packages/tooltips/src/elements/Tooltip.tsx:19]: The component author will need to take care with this ☝️ type of reuse, as `undefined` can come through from the imported interface in a manner that makes the local styled interface feel less expressive. In this case, it could appear that `reusedProp` is mandatory. But if `ITooltipProps['prop']` is optional, then `reusedProp` may also be `undefined`.

tldr; I'm not entirely convinced this is an ideal reuse pattern.

Comment (jzempel) [reply packages/tooltips/src/elements/Tooltip.tsx:19]: fwiw, there are [TypeScript utilities](https://www.typescriptlang.org/docs/handbook/utility-types.html) to help with this. It is just a matter of the author being meticulous. For example:

```ts
import { ITooltipProps } from '...';

interface IStyledTooltipProps {
  reusedProp: NonNullable<ITooltipProps['prop']>;
}
```

File: packages/tooltips/src/styled/StyledTooltip.ts [hunk 6]
```diff
@@ -13,23 +13,17 @@ import {
...
 
 const COMPONENT_ID = 'tooltip.tooltip';
 
-export type TOOLTIP_SIZE = 'small' | 'medium' | 'large' | 'extra-large';
-export type TOOLTIP_TYPE = 'light' | 'dark';
-
-export interface IStyledTooltipProps {
-  hasArrow?: boolean;
-  size?: TOOLTIP_SIZE;
+export interface IStyledTooltipProps extends Pick<ITooltipProps, 'hasArrow' | 'size' | 'zIndex'> {
```
Comment (jzempel) [inline packages/tooltips/src/styled/StyledTooltip.ts:23]: After some additional noodling, I think this is a fairly decent pattern for public interface reuse within view components.

File: packages/tooltips/src/elements/Tooltip.spec.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Francois-Esquire) [reply packages/tooltips/src/elements/Tooltip.spec.tsx:13]: I have been thinking about this since we first discussed this. I'm actually in favor of `types/` since it felt like `types` shouldn't categorically fit under `utils/`. Another reason why: this would make  it easier for `utils/`, `elements/` and `styles/` to `import '../types;'` instead of adding `import '../utils/types` (for `elements` and `styles`).

File: packages/tooltips/src/styled/StyledTooltip.ts [hunk 7]
```diff
@@ -13,23 +13,17 @@ import {
...
   /** All valid [Popper.JS Placements](https://popper.js.org/popper-documentation.html#Popper.placements) */
-  placement: POPPER_PLACEMENT;
-  zIndex?: number | string;
-  type: TOOLTIP_TYPE;
+  placement: typeof POPPER_PLACEMENT[number];
+  type: NonNullable<ITooltipProps['type']>;
 }
```
Comment (Francois-Esquire) [inline packages/tooltips/src/styled/StyledTooltip.ts:27]: Should we move `styled/` types under the same umbrella? Or is it strictly for the public API?
If so, maybe a `types/styled.ts` would be a good way of organizing it without necessarily exporting it.

File: packages/tooltips/src/utils/types.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+export interface ITooltipProps extends HTMLAttributes<HTMLDivElement> {
```
[the remark this answers is not in the document]
Comment (Francois-Esquire) [reply packages/tooltips/src/utils/types.ts:49]: I have been thinking about this as well: we can post process the results and iterate through the props per component, however for documentations sake, there are some external APIs throughout `react-components` (`react-popper` and `downshift`) that we can annotate the API with as fragments - like what we do for `dropdowns`.

File: packages/tooltips/src/elements/Tooltip.spec.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jzempel) [reply packages/tooltips/src/elements/Tooltip.spec.tsx:13]: Ok, I like it. Thanks for the feedback.

File: packages/tooltips/src/utils/types.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jzempel) [reply packages/tooltips/src/utils/types.ts:49]: Ok, I'm happy to see how this works out in practice. In the meantime, I think the description with corresponding reference link is a solid compromise here.

File: packages/tooltips/src/styled/StyledTooltip.ts (hunk 7 above)
Comment (jzempel) [reply packages/tooltips/src/styled/StyledTooltip.ts:27]: I'm opposed to re-situating these private interfaces away from their associated styled components to a central location. I feel like it communicates the wrong thing re: visibility – but that's just my personal preference and I'm happy to defer to the team's direction on this. To your point, we might discover some nice type affinity that could be gained if these are co-located 🤷.

Either way, my current tooling exercise is not going to engage with the `styled` interfaces. That's a substantial effort; my plan is to stay focused on the publicly exported element component interfaces.

**Review by hzhu — APPROVED** [review]
